### PR TITLE
REF: Pieces broken off of #24024

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @TomAugspurger 

If I did this right, this shouldn't cause rebasing headaches.  LMK if I messed up on that and I'll try to fix.

Nearly everything here is verbatim from #24024.  I'll comment inline with pieces that are not.

### CODE REVIEW

File: pandas/core/arrays/period.py [hunk 1]
```diff
@@ -983,7 +983,7 @@ def dt64arr_to_periodarr(data, freq, tz=None):
 
     """
     if data.dtype != np.dtype('M8[ns]'):
-        raise ValueError('Wrong dtype: %s' % data.dtype)
+        raise ValueError('Wrong dtype: {dtype}'.format(dtype=data.dtype))
```
Comment (jbrockmendel) [inline pandas/core/arrays/period.py]: Not from #24024, but @jreback  commented there asking for this change.

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -227,11 +227,11 @@ def __new__(cls, data=None,
...
-            result = cls._generate_range(start, end, periods,
-                                         freq=freq, tz=tz, normalize=normalize,
-                                         closed=closed, ambiguous=ambiguous)
-            result.name = name
-            return result
+            dtarr = DatetimeArray._generate_range(
+                start, end, periods,
+                freq=freq, tz=tz, normalize=normalize,
+                closed=closed, ambiguous=ambiguous)
+            return cls(dtarr, name=name)
```
Comment (jbrockmendel) [inline pandas/core/indexes/datetimes.py]: I think in #24024 this becomes `cls._simple_new(dtarr, name=name)`.  That isn't possible yet here.

Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: ditto below and for TimedeltaIndex

File: pandas/tests/arrays/test_timedeltas.py [hunk 3]
```diff
@@ -8,6 +8,17 @@
...
 
+class TestTimedeltaArrayConstructor(object):
+    def test_copy(self):
+        data = np.array([1, 2, 3], dtype='m8[ns]')
+        arr = TimedeltaArray(data, copy=False)
+        assert arr._data is data
+
+        arr = TimedeltaArray(data, copy=True)
+        assert arr._data is not data
+        assert arr._data.base is not data
```
Comment (jbrockmendel) [inline pandas/tests/arrays/test_timedeltas.py]: This last line is not present in the test in #24024.

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 4]
```diff
@@ -246,6 +247,17 @@ def test_to_datetime_parse_timezone_keeps_name(self):
 
 
 class TestToDatetime(object):
+    def test_to_datetime_dtarr(self):
```
Comment (jbrockmendel) [inline pandas/tests/indexes/datetimes/test_tools.py]: This test is not in #24024, but this tests the changes in core.tools.datetimes

**Comment by jbrockmendel** [comment]
In terms of LOC this shaves about 5% off of 24024.  Not huge, but along with searchsorted, repeat, and reductions getting broken off, these add up

**Comment by TomAugspurger** [comment]
#24024 doesn't apply cleanly on top of this.

Could we instead give a +1 to the changes here, and then you make a PR against my branch for #24024 with the new changes?

File: pandas/core/indexes/datetimes.py (hunk 2 above)
Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: Is it worth doing this now then, when we'll just need to change it again?

File: pandas/core/arrays/datetimelike.py [hunk 5]
```diff
@@ -451,7 +464,7 @@ def _isnan(self):
         return (self.asi8 == iNaT)
 
     @property  # NB: override with cache_readonly in immutable subclasses
-    def hasnans(self):
+    def _hasnans(self):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: IIRC this is a general Index property, are changing this globally? (ok by me)

**Review by jreback — CHANGES_REQUESTED** [review]
i would split out the hasnans change and do that separately as it needs to happen globally

```
(bamboo-dev) jreback@dev:~/pandas-dev$ grep -r hasnans pandas --include '*.py'|cut -d ' ' -f 1|uniq
pandas/core/arrays/timedeltas.py:
pandas/core/arrays/period.py:
pandas/core/arrays/datetimelike.py:
pandas/core/arrays/datetimes.py:
pandas/core/indexes/timedeltas.py:
pandas/core/indexes/interval.py:
pandas/core/indexes/category.py:
pandas/core/indexes/numeric.py:
pandas/core/indexes/datetimelike.py:
pandas/core/indexes/base.py:
pandas/core/dtypes/dtypes.py:
pandas/core/series.py:
pandas/core/resample.py:
pandas/core/base.py:
pandas/tests/test_lib.py:
pandas/tests/indexes/datetimes/test_ops.py:
pandas/tests/indexes/interval/test_interval.py:
pandas/tests/indexes/period/test_ops.py:
pandas/tests/indexes/multi/test_missing.py:
pandas/tests/indexes/multi/test_missing.py:def
pandas/tests/indexes/multi/test_missing.py:
pandas/tests/indexes/common.py:
pandas/tests/indexes/timedeltas/test_ops.py:
pandas/tests/indexes/test_common.py:
pandas/tests/series/test_internals.py:def
pandas/tests/series/test_internals.py:
```

**Comment by TomAugspurger** [comment]
What do you mean by globally changing the hasnans? I don't think anyone was
suggesting changing that anywhere other than on the datelike Array classes,
where we (I) don't want them as part of the public API.

**Comment by jbrockmendel** [comment]
> I don't think anyone was suggesting changing that anywhere other than on the datelike Array classes,

This matches what I had in mind, just doing it on DTI/TDI/PI.

**Comment by jreback** [comment]
> What do you mean by globally changing the hasnans? I don't think anyone was
suggesting changing that anywhere other than on the datelike Array classes,
where we (I) don't want them as part of the public API.

that's fine, but wont' accept a change that also does not fix the existing Index codebases. This is SO confusing you need as much consistency as possible.

**Comment by jbrockmendel** [comment]
Did we ever reach a consensus here?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks. I realized you are just changing the non-Index idiom here which is in-line with other idioms.

**Pull request merged**